File: Solution/CSharp/338.CountingBits.cs
static public int[] CountBits(int n) {
    int[] arr = new int[n + 1];
    arr[0] = 0;

    int pow2 = 2;
    int before = 1;
    for (int i = 1; i < n + 1; i++) {
        if (i == pow2) {
            arr[i] = 1;
            before = 1;
            pow2 <<= 1;
        } else {
            arr[i] = arr[before] + 1;
            before++;
        }
    }

    return arr;
}

File: Solution/Python/338. Counting Bits.py
# dp
class Solution:
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]
        """
        res = [0] * (num + 1)

        pow2 = 2
        before = 1
        for i in range(1, num + 1):
            if i == pow2:
                res[i] = 1
                before = 1
                pow2 *= 2
            else:
                res[i] = res[before] + 1
                before += 1

        return res


s = Solution()
num = 8
res = s.countBits(num)
print(res)
